### ISSUE TITLE
Possible False Positibe MA0055

### ISSUE THREAD

**Issue by DAud-IcI** [meziantou/Meziantou.Analyzer#286]
**Context**
Rule identifier: _MA0055_
Target Framework: _.NET 5.0_
Language version: _9.0_

**Describe the bug**
I'm releasing my unmanaged resources via IDisposable, and dispose is called from the finalizer as a fallback. Is that not a valid scenario?

**To Reproduce**
Code or steps to reproduce the behavior:

````c#
public class MyClass : IDisposable
{
    ~MyClass() => Dispose(disposing: false);

    public Dispose()
    {
        Dispose(disposing: false);
        GC.SuppressFinalize(this);
    }

    protected virtual void Dispose(bool disposing)
    {
        if (disposing) { /* Release managed. */ }
        // Release unmanaged.
    }
}
````

**Additional context**

**Comment by meziantou** [comment]
Finalizers should be used to ensure native resources are released. Since .NET 2.0, you should better encapsulate the native resources into a `SafeHandle` when possible. This way you don't have to deal with the complexity of finalizers.

[Stop using IntPtr for dealing with system handles](https://www.meziantou.net/stop-using-intptr-for-dealing-with-system-handles.htm)

**Comment by Piedone** [comment]
The issue here is that the class in the example doesn't necessarily has to do anything with unmanaged resources, especially directly. Implementing `IDisposable` doesn't necessarily warrant using `SafeHandle`. The same would be an issue for one that does use unmanaged resources but indirectly (like using a `Bitmap` or `TextWriter`). I don't see why it would make things better to encapsulate those into a `SafeHandle`.

So using `SafeHandle` is good if you're interacting with unmanages resources directly, like with P/Invoke, otherwise, it's up to the other implementation.

**Comment by meziantou** [comment]
For the context, I've never written a finalizer in my applications as I can't see a usage for them. But I hope you'll prove me I'm wrong :) I don't say it's useless, but it seems more like a safeguard to clean up resources in the case `Dispose` failed. There is no guarantee the finalizer is called in the application lifetime (only if GC is needed).

For specific examples:
- Bitmap: The Image class already has a finalizer. Why do you need another one in your own class?
- TextWriter: This class doesn't contain any unmanaged resources. Why would you need a finalizer in this case?

**Comment by Piedone** [comment]
A finalizer is indeed just a backup in all these cases if the consumer doesn't call `Dispose()`. And the only thing it does is just call `Dispose()`, including for the objects it owns, like a `Bitmap` or `TextWriter`.

**Comment by meziantou** [comment]
> **including for the objects it owns**, like a Bitmap or TextWriter

From my understanding, it shouldn't if these objects already have finalizers (such as `Bitmap` from its base class `Image`). Also, I don't understand why you talk about `TextWriter` as this class doesn't hold any native resource.

Maybe you can provide a real-world example of a finalizer where it is useful and SafeHandle cannot be used? Is it a common use-case or an edge case? For edge cases, I think it's better to disable the rule using the `SuppressMessage` attribute.

**Comment by Piedone** [comment]
I'm mentioning `TextWriter` just as an example of an `IDisposable`.

That being said, the sample under https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1063 does actually only recommend a finalizer for only unmanaged resources. So maybe we just shouldn't have a finalizer in such case. (And on a second look we do have other `IDispsable`s that directly own unmanaged resources as well where `SafeHandle` should be used.)

**Comment by meziantou** [comment]
Thanks for pointing CA1063. This confirms that only classes that directly own unmanaged resources should have a finalizer. Also, the [finalizer documentation](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/finalizers) states that in most cases, you can avoid writing a finalizer by using the SafeHandle or derived classes to wrap any unmanaged handle.
So, very few classes should have a finalizer. I think this is acceptable to use the `SuppressMessage` attribute for edge cases and keep the analyzer as-is. (maybe the documentation could be improved to give more guidance)

**Comment by Piedone** [comment]
Thanks, this is not an issue then. @DAud-IcI please close.
